File: lib/pymodules/cmd_misc.py
'''
cmd_misc.c

a collection of miscellaneous commands that come with NakedMud(tm)
'''
import mudsys, mud, hooks, event, mudsys



def cmd_stop(ch, cmd, arg):
    '''If you are currently performing an action (for example, a delayed
       command), make an attempt to stop performing that action.'''
    if not ch.isActing():
        ch.send("But you're not currently performing an action!\r\n")
    else:
        ch.interrupt()

def cmd_clear(ch, cmd, arg):
    '''This command will clear your display screen.'''
    ch.send_raw("\033[H\033[J")

def event_delayed_cmd(ch, filler, cmd):
    '''used to perform delayed commands'''
    ch.act(cmd, True)

def cmd_delay(ch, cmd, arg):
    '''Usage: delay <seconds> <command>

       Allows the user to prepare a command to be executed in the future. For
       example:

       > delay 2 say hello, world!

       Will make you say \'hello, world!\' two seconds after entering the
       delayed command.'''
    try:
        secs, to_delay = mud.parse_args(ch, True, cmd, arg, "double string")
    except: return

    if secs < 1:
        ch.send("You can only delay commands for positive amounts of time.")
    else:
        ch.send("You delay '%s' for %.2f seconds" % (to_delay, secs))
        event.start_event(ch, secs, event_delayed_cmd, None, to_delay)

def cmd_motd(ch, cmd, arg):
    '''This command will display to you the mud\'s message of the day.'''
    ch.page(get_motd())

def cmd_save(ch, cmd, arg):
    '''Attempt to save your character and all recent changes made to it, to
       disk. This automatically happens when logging out.'''
    if mudsys.do_save(ch):
        ch.send("Saved.")
    else:
        ch.send("Your character was not saved.")
    

def cmd_quit(ch, cmd, arg):
    '''Attempts to save and log out of the game.'''
    mud.log_string(ch.name + " has left the game.")
    mudsys.do_save(ch)
    mudsys.do_quit(ch)



################################################################################
# add our commands
################################################################################
mudsys.add_cmd("stop",  None, cmd_stop,  "player", False)
mudsys.add_cmd("clear", None, cmd_clear, "player", False)
mudsys.add_cmd("delay", None, cmd_delay, "player", False)
mudsys.add_cmd("motd",  None, cmd_motd,  "player", False)
mudsys.add_cmd("save",  None, cmd_save,  "player", False)
mudsys.add_cmd("quit",  None, cmd_quit,  "player", True)

chk_can_save = lambda ch, cmd: not ch.is_npc
mudsys.add_cmd_check("save", chk_can_save)
mudsys.add_cmd_check("quit", chk_can_save)

File: src/cmd_misc.c
//*****************************************************************************
//
// cmd_misc.c
//
// a collection of miscellaneous commands that come with NakedMud(tm)
//
//*****************************************************************************
#include "mud.h"
#include "character.h"
#include "socket.h"
#include "utils.h"
#include "save.h"
#include "event.h"
#include "action.h"
#include "handler.h"


//
// stop performing the character's current action
//
COMMAND(cmd_stop) {
#ifdef MODULE_FACULTY
  if(!is_acting(ch, FACULTY_ALL))
    send_to_char(ch, "But you're not currently performing an action!\r\n");
  else
    interrupt_action(ch, FACULTY_ALL);
#else
  if(!is_acting(ch, 1))
    send_to_char(ch, "But you're not currently performing an action!\r\n");
  else
    interrupt_action(ch, 1);
#endif
}


//
// clear the screen
//
COMMAND(cmd_clear) {
  send_to_char(ch, "\033[H\033[J");
}


//
// quit the game
//
COMMAND(cmd_quit)
{
  char buf[MAX_BUFFER];

  /* log the attempt */
  sprintf(buf, "%s has left the game.", charGetName(ch));
  log_string(buf);

  save_player(ch);

  // 
  // gotta make sure we have a socket. Who knows...
  // a mobile might be trying to quit
  //
  if(charGetSocket(ch)) {
    SOCKET_DATA *sock = charGetSocket(ch);
    charSetSocket(ch, NULL);
    socketSetChar(sock, NULL);
    close_socket(sock, FALSE);
  }

  extract_mobile(ch);
}


//
// save the character
//
COMMAND(cmd_save)
{
  save_player(ch);
  text_to_char(ch, "Saved.\r\n");
}


//
// the function for executing a delayed command
//
void event_delayed_cmd(CHAR_DATA *ch, void *data, char *cmd) {
  do_cmd(ch, cmd, TRUE, TRUE);
}


//
// Perform a command, but delay its execution by a couple seconds
//
COMMAND(cmd_delay) {
  if(!arg || !*arg) {
    send_to_char(ch, "What did you want to delay, and by how long?\r\n");
    return;
  }

  // how long should we delay, in seconds?
  char time[SMALL_BUFFER];
  arg = one_arg(arg, time);

  // no command to delay
  if(!*arg) {
    send_to_char(ch, "What command did you want to delay?\r\n");
    return;
  }
  if(atoi(time) < 1) {
    send_to_char(ch, "You can only delay commands for positive amounts of time.\r\n");
    return;
  }

  send_to_char(ch, "You delay '%s' for %d seconds.\r\n", arg, atoi(time));
  start_event(ch, atoi(time) SECONDS, event_delayed_cmd, NULL, NULL, arg);
}


//
// An entrypoint into the character's notepad
//
/*
COMMAND(cmd_write) {
  if(!charGetSocket(ch))
    send_to_char(ch, "You cannot use notepad if you have no socket.\r\n");
  else {
    send_around_char(ch, TRUE, "%s pulls out a notepad and begins writing.", 
		     charGetName(ch));
    start_notepad(charGetSocket(ch), "");
  }
}
*/
